ui/src/criteria/concept.tsx

```diff
@@ -163,39 +177,40 @@ function ConceptEdit(props: ConceptEditProps) {
     )
   );
 
-  const updater = useCohortUpdater();
+  const dispatch = useAppDispatch();
 
   return (
     <Loading status={conceptsState}>
       {conceptsState.data ? (
         <TreeGrid
           columns={allColumns}
           data={conceptsState.data}
-          prefixElements={(id: TreeGridId, data: TreeGridRowData) => {
-            const index = props.criteria.selected.findIndex(
-              (row) => row.id === id
-            );
+          prefixElements={(id: TreeGridId, rowData: TreeGridRowData) => {
+            const index = props.data.selected.findIndex((row) => row.id === id);
 
             return (
               <Checkbox
                 size="small"
                 checked={index > -1}
                 inputProps={{ "aria-label": "controlled" }}
                 onChange={() => {
-                  updater.updateCriteria(
-                    props.group.id,
-                    props.criteria.id,
-                    (criteria: ConceptCriteria) => {
-                      if (index > -1) {
-                        criteria.selected.splice(index, 1);
-                      } else {
-                        const name = data["concept_name"];
-                        criteria.selected.push({
-                          id: id as number,
-                          name: !!name ? String(name) : "",
-                        });
-                      }
-                    }
+                  dispatch(
+                    updateCriteriaData({
+                      cohortId: props.cohort.id,
+                      groupId: props.group.id,
+                      criteriaId: props.criteriaId,
+                      data: produce(props.data, (data) => {
```

Comment:
I'm not quite able to follow what's happening here, mainly due to a lack of redux experience. Can you explain why immer's produce is necessary here?

Reply:
I can't change the existing data, just like with React state. So I would otherwise have to deep copy data or use spreads to get a new object. immer handles that automatically and allows more natural code using methods like splice.

Reply:
The end game here is to stop using the generic updateData method and instead allow the plugins to implement their own actions and reducer. Then this would change to two actions: select and unselect, and the parameters would just be the ids so there would be no need to copy the data here. That will also be much nicer for undo/redo messages. This is already a large change though so I tried to keep things the same where possible.

Reply:
I see, thanks. Sorry - I missed that this was modifying `props.data`, and this was not a normal redux update (which from reading the docs - it seems already uses immer behind the scenes).